src/stable/build/commands/build/wasm_binary/rust/stable_canister_template/src/ic/call_raw.rs

```diff
@@ -48,95 +44,199 @@ pub fn get_function(ctx: Ctx) -> QuickJsResult<Function> {
-                let _cleanup = scopeguard::guard((), |_| {
-                    let result = cleanup(ctx.clone(), &global_resolve_id, &global_reject_id);
+                let _cleanup_scopeguard = scopeguard::guard((), |_| {
+                    let result = with_ctx(|ctx| {
+                        // JavaScript code execution: macrotask when running as an ICP cleanup callback
+                        // The ICP cleanup callback is executed when an inter-canister call reply or reject callback traps
+                        // If the reply or reject callback does not trap, this will run in the same call
+                        // as the reply or reject callback, and thus won't be a macrotask, but will still
+                        // be a JavaScript execution. If the inter-canister call returns synchronously because
+                        // of a synchronous error, this will run in the same ICP call as the original
+                        // execution that called into call_raw (call in JavaScript), and thus won't be a JavaScript macrotask
+                        // but will still be a JavaScript execution
+                        cleanup(
+                            ctx.clone(),
+                            &global_resolve_id,
+                            &global_reject_id,
+                            is_recovering_from_trap(),
+                        )?;
+
+                        // We must drain all microtasks that could have been queued during previous JavaScript executions
+                        // Those executions include the settle_promise macrotask, the cleanup call above as a regular
+                        // JavaScript execution, or the cleanup callback above as a macrotask execution
+                        drain_microtasks(&ctx);
+
+                        Ok(())
+                    });
 
                     if let Err(e) = result {
                         trap(&format!("Azle CallRawCleanupError: {e}"));
                     }
+
+                    // We must drain all inter-canister call futures that could have been queued during previous JavaScript executions
+                    // Those executions include the settle_promise macrotask, the cleanup call above as a regular
+                    // JavaScript execution, or the cleanup callback above as a macrotask execution
+                    // This MUST be called outside of the with_ctx closure
+                    drain_inter_canister_call_futures();
                 });
 
                 let call_result = call_raw128(canister_id, &method, args_raw, payment).await;
 
-                let result = resolve_or_reject(
-                    ctx.clone(),
-                    &call_result,
-                    &global_resolve_id,
-                    &global_reject_id,
-                );
+                let result = with_ctx(|ctx| {
+                    // JavaScript code execution: macrotask
```

Comment:
Reanalyze the macrotask comments in here, we should explain when it is and when it is not a macrotask